### DISCUSSION TITLE
Adding option to set zlib/deflate compression level

### DISCUSSION THREAD

**Discussion by mnmtz** [py-pdf/pypdf#1929] [Ideas]
Adding an option to set zlib/deflate compression level from 1-9 (lowest to highest compression) for compressing content streams.

**Comment by pubpub-zz** [comment]
currently pypdf is using default compression (6).
can you please explain why you would like to change the compression level.
These two articles show no major gain when changing compression level

https://stackoverflow.com/questions/76001725/how-can-i-improve-the-pdf-compression-quality-in-my-python-code-using-the-pymupd
https://github.com/zlib-ng/zlib-ng/discussions/871

**Comment by mnmtz** [comment]
I am currently working on a project, where every kB in the final PDF matters and even if there is only a small reduction in size, it would help.

**Comment by pubpub-zz** [comment]
I order to motivate can you do some test adjuing directly the values in the code to show the benefit on a set of pdf files

**Reply by MartinThoma** [reply]
As a hint @mnmtz : 

* https://github.com/py-pdf/sample-files/blob/main/009-pdflatex-geotopo/GeoTopo.pdf
* https://github.com/py-pdf/pypdf/blob/main/pypdf/filters.py#L237 - see https://docs.python.org/3/library/zlib.html#zlib.compress
* you can install a local pypdf version with `pip install -e .` where the dot is the local directory.

You could create a table of `level | resulting file size` to show your point that it's worth it :-)

**Reply by mnmtz** [reply]
Thank you very much! I have managed to set the level manually so there is no need to create a dedicated parameter if not requested otherwise.

**Reply by MartinThoma** [reply]
We are trying to make you become a contributor :smile: Please share your findings with us: What was the original file size and what was the size with the different compression levels?

**Comment by mnmtz** [comment]
Here are some examples with different compression levels:

[compression_levels.csv](https://github.com/py-pdf/pypdf/files/12068245/compression_levels.csv)
